lib/_http_outgoing.js

```diff
@@ -526,7 +527,14 @@ OutgoingMessage.prototype.end = function(data, encoding, callback) {
 
   var self = this;
   function finish() {
-    self.emit('finish');
+    // NOTE: It is very important to wait for `prefinish` event.
+    // It is emitted only when the response becomes active in the pipeline.
+    // The ordering of requests/responses in `_http_server.js` is dependent
+    // on this, doing it in different order will result in crash.
+    if (self._prefinish)
+      self.emit('finish');
+    else
+      self.on('prefinish', finish);
```

Comment:
It is indeed very confusing because we don't use streams API for these events. We should move to them in the future. Feel free to open issue for this, @trevnorris